### ISSUE TITLE
ods files unzipped

### ISSUE THREAD

**Issue by pcav** [3liz/qgis-wfsOutputExtension#11]
Apparently the exported ods file is unzipped, so it cannot be opened directly by LibreOffice.

**Comment by Gustry** [comment]
Sorry @pcav I don't understand the issue. Libreoffice wants to read a plain ODS file, not a zipped ODS. Why do you want a zipped ODS?

Maybe I'm missing something?

**Comment by rldhont** [comment]
ODS format is a zip file that contains an XML and other data like images.

The ODS file is generated by QGIS QgsVectorFileWriter.

**Comment by Gustry** [comment]
Yes, but I tried on Lizmap, I have an ODS file and it's working well in LibreOffice. I don't have any zipped file.

Are you sure it wasn't your file explorer which opening the file with the archive software?

**Comment by pcav** [comment]
so you can open an unzipped ODS file from LibreOffice? In my case it opened as xml.

**Comment by Gustry** [comment]
I'm not "unzipping" the ODS file, everything works fine.

* Lizmap (WfsOutputExtension) gives me an `my_layer.ods` file when I select `ODS`.
* Firefox downloads it as ODS file.
* If I double click on it (the ODS file), LibreOffice opens normally, loading the ODS file.
* I also tried to open LibreOffice, open the ODS file from LibreOffice, everything is fine.
* If I right click on the file, choose `Open with` my archive software, yes indeed I can see XML files inside. But you shouldn't to that. It's like opening a QGZ file with your archive software.

@pcav Does it happen with other ODS files?

**Comment by pcav** [comment]
OK, your explanation makes sense to me. The problem stems from the fact that the browser by default suggests to open the ods with LO.
Thanks.

**Comment by Gustry** [comment]
Ok, so I just checked the MIME Type on the file:

```
~/dev/qgis/server_plugin $ file --mime-type -b /home/etienne/Téléchargements/export_value_map.ods 
application/zip
```
So indeed, your webbrowser is following the Mime Type to choose the archive software.
My firefox is following the extension `.ods` and doesn't open my archive software.

The content type in the plugin is well defined:
* https://github.com/3liz/qgis-wfsOutputExtension/blob/master/wfsOutputExtension/wfs_filter.py#L70
* https://github.com/3liz/qgis-wfsOutputExtension/blob/master/wfsOutputExtension/wfs_filter.py#L145

I didn't check if QGIS Server can set the correct MIME Type. Maybe @dmarteau has an idea?

**Comment by Gustry** [comment]
> Ok, so I just checked the MIME Type on the file:
> 
> ```
> ~/dev/qgis/server_plugin $ file --mime-type -b /home/etienne/Téléchargements/export_value_map.ods 
> application/zip
> ```

And on a ODS file from scratch created in LO:

```bash
~/dev/qgis/server_plugin $ file --mime-type -b /tmp/test_ods.ods 
application/vnd.oasis.opendocument.spreadsheet
```

**Comment by Gustry** [comment]
if I'm not wrong, this has been fixed ? (with #16)

Gentle ping @pcav

**Comment by u-cav** [comment]
@Gustry the issue seems to be fixed on the user side. Firefox offers to open the file with LO, and the downloaded file can be opened directly.

The mimetype of the downloaded file is still:

`application/zip`

instead of:

`application/vnd.oasis.opendocument.spreadsheet`

**Comment by rldhont** [comment]
In the response headers provided by QGIS Server or Lizmap you would find `application/vnd.oasis.opendocument.spreadsheet` for content type. But between QGIS Server / Lizmap and the Browser there is some software that probably not recognize `application/vnd.oasis.opendocument.spreadsheet` content type and change it to `application/zip` because ODS is a zip file.

**Comment by u-cav** [comment]
@rldhont I would consider it fixed, since the behaviour is correct on the user end. Thanks!

**Comment by Gustry** [comment]
The MIME type is still indeed incorrect, but for the web browser, it seems alright indeed.
